Improve Firestore delete

samtstern opened the pull request:
<!--

Thank you for contributing to the Firebase community! Please fill out the form below.

Run the linter and test suite
==============================
Run `npm test` to make sure your changes compile properly and the tests all pass on your local machine. We've hooked up this repo with continuous integration to double check those things for you.

-->

### Description

Started with an investigation into large document deletes, made a few changes:

- Move the "Transaction too big" error handling code into the operation so that it can be handled visibly
- Before doing n parallel delete batches wait until one succeeds. For cases where all documents are large this means we'll get the right batch size sooner.
- Fix some logging.

TODO:

- [x] Merge #2940
- [x] Address https://github.com/firebase/firebase-tools/pull/2940#discussion_r543714192

### Scenarios Tested

Wrote a small script to add 1000 big docs and 1000 small docs to collections:

```
$ time node index.js 
Adding 1000 small documents
Done
Adding 1000 big documents
Done

real    0m49.459s
user    0m6.877s
sys     0m0.842s
```

Deleting the small docs works as always:

```
$ firebase --project=fir-dumpster firestore:delete --recursive --yes small-data
Deleted 250 docs (250000 docs/s)
Deleted 500 docs (810 docs/s)
Deleted 750 docs (934 docs/s)
Deleted 1000 docs (1037 docs/s)
```

Deleting the big docs hits the new error handling and succeeds (slowly):

```
$ firebase --project=fir-dumpster firestore:delete --recursive --yes big-data 
⚠  firestore: delete transaction too large, reducing batch size from 250 to 25
Deleted 25 docs (Infinity docs/s)
Deleted 50 docs (8 docs/s)
Deleted 75 docs (11 docs/s)
Deleted 100 docs (15 docs/s)
Deleted 125 docs (18 docs/s)
Deleted 150 docs (21 docs/s)
Deleted 175 docs (24 docs/s)
Deleted 200 docs (27 docs/s)
Deleted 225 docs (29 docs/s)
Deleted 250 docs (32 docs/s)
Deleted 275 docs (34 docs/s)
Deleted 300 docs (37 docs/s)
Deleted 325 docs (39 docs/s)
Deleted 350 docs (42 docs/s)
Deleted 375 docs (39 docs/s)
Deleted 400 docs (39 docs/s)
Deleted 425 docs (39 docs/s)
Deleted 450 docs (35 docs/s)
Deleted 475 docs (37 docs/s)
Deleted 500 docs (35 docs/s)
Deleted 525 docs (37 docs/s)
Deleted 550 docs (37 docs/s)
Deleted 575 docs (39 docs/s)
Deleted 600 docs (39 docs/s)
Deleted 625 docs (40 docs/s)
Deleted 675 docs (42 docs/s)
Deleted 725 docs (45 docs/s)
Deleted 750 docs (46 docs/s)
Deleted 775 docs (47 docs/s)
Deleted 800 docs (47 docs/s)
Deleted 825 docs (48 docs/s)
Deleted 850 docs (37 docs/s)
Deleted 875 docs (37 docs/s)
Deleted 900 docs (38 docs/s)
Deleted 925 docs (38 docs/s)
Deleted 975 docs (40 docs/s)
Deleted 1000 docs (41 docs/s)
```


### Sample Commands

N/A

samtstern commented:
@bkendall I made as few changes to the file as possible when TS-ifying but unfortunately GitHub still sees it as all new.

samtstern commented on src/firestore/delete.ts:
```diff
@@ -0,0 +1,539 @@
+        })
+        .catch((e: any) => {
+          // If the transaction is too large, reduce the batch size
+          if (
```
Here's the main new bit of error handling.

samtstern commented on src/firestore/delete.ts:
```diff
@@ -0,0 +1,539 @@
+
+      // We want to see one batch succeed before we scale up, so this case
+      // limits parallelism until first success
+      if (numDocsDeleted === 0 && numPendingDeletes >= 1) {
```
This condition is also new.

yuchenshi commented:
Sorry, no bandwidth reviewing this, but would you mind split the TS-ify out of the PR (at least into its own commit) so it's easier to spot the actual functional changes? Right now, the PR is huge and it's hard to tell the real difference.

samtstern commented:
@yuchenshi yeah I wish I had thought of that.  I'll see what I can do.

samtstern commented:
Actually as @yuchenshi suggested going to move the TS-ification to a separate PR:
https://github.com/firebase/firebase-tools/pull/2940

bkendall commented on src/firestore/delete.ts:
```diff
@@ -248,23 +274,17 @@ export class FirestoreDelete {
-            return x.document;
-          });
-      });
+    return this.apiClient.post("/v1/" + url, body).then((res) => {
```
You can provide `apiVersion: "v1"` to the Client constructor above, if you wish.

Also here to say: this is an intentional change to `v1`, right?

samtstern replied:
Yes this is intentional, and good to know!

bkendall commented on src/firestore/delete.ts:
```diff
@@ -248,23 +274,17 @@ export class FirestoreDelete {
+    return this.apiClient.post("/v1/" + url, body).then((res) => {
+      // Return the 'document' property for each element in the response,
+      // where it exists.
+      return (res.body as any[])
```
The `post` method allows for templating so you don't have to do this. The first is the request body type, the second is the response body type (maybe this is too specific? attempting to show an array of objects w/ documents):

```typescript
this.apiClient.post<any, Array<{ document: any }>>(...)
```

samtstern replied:
Nice!

bkendall commented on src/firestore/delete.ts:
```diff
@@ -348,11 +375,37 @@ export class FirestoreDelete {
+          // If the transaction is too large, reduce the batch size
+          if (
+            e.status === 400 &&
+            e.message.indexOf("Transaction too big") !== -1 &&
```
```suggestion
            e.message.includes("Transaction too big") &&
```

bkendall commented on src/gcp/firestore.ts:
```diff
@@ -60,26 +60,10 @@ export async function deleteDocuments(project: string, docs: any[]): Promise<num
-    }
-    throw err;
-  }
+  const res = await api.request("POST", url, {
```
I guess this isn't a requirement, but it would be nice if `apiv2` was used in this file while you're in here :) (I see this isn't really rewriting anything though, so I get if you don't want to heh)

samtstern replied:
For you?  Anything.

bkendall commented on CHANGELOG.md:
```diff
@@ -0,0 +1 @@
+- Better error handling behavior for `firestore:delete` when deleting very large documents.
```
Suggestion: ("Improves" matches the desired schema for these entries)

```suggestion
- Improves error handling for `firestore:delete` when deleting very large documents.
```

bkendall commented on src/firestore/delete.ts:
```diff
@@ -1,18 +1,20 @@
 import { FirebaseError } from "../error";
 import * as logger from "../logger";
 import * as utils from "../utils";
+import { firestoreOriginOrEmulator } from "../api";
```
👍  this is appreciated

bkendall reviewed (approved):
approval with a few little touches. looks good though.

Would you want to add a similar block to the [client integration tests](https://github.com/firebase/firebase-tools/blob/master/scripts/client-integration-tests/tests.ts#L93-L115) for testing Firestore actions? Having a decent practical test could give us good coverage in our logic for Firestore

samtstern commented:
@bkendall adding a client integration test for this isn't really feasible (unfortunately) since there is no `firestore:set` operation to create data which we could then delete with `firestore:delete`

samtstern commented:
Also tested deleting a DB with mixed big and small documents:
```
$ firebase --project=fir-dumpster firestore:delete --recursive --all-collections? You are about to delete THE ENTIRE DATABASE for fir-dumpster. Are you sure? Yes
Deleting the following collections: big-data-batch, big-data-bulk, small-data, small-data-batch, small-data-bulk
Deleted 250 docs (Infinity docs/s)
Deleted 500 docs (3448 docs/s)
Deleted 750 docs (4601 docs/s)
Deleted 1000 docs (1126 docs/s)
Deleted 1250 docs (1310 docs/s)
Deleted 1500 docs (1478 docs/s)
Deleted 2250 docs (2117 docs/s)
Deleted 2500 docs (1998 docs/s)
Deleted 2750 docs (2075 docs/s)
Deleted 2801 docs (2024 docs/s)
Deleted 3252 docs (2268 docs/s)
Deleted 3502 docs (2283 docs/s)
Deleted 4252 docs (2693 docs/s)
Deleted 5002 docs (3052 docs/s)
Deleted 5502 docs (3227 docs/s)
Deleted 5752 docs (3160 docs/s)
Deleted 6002 docs (3193 docs/s)
Deleted 6252 docs (2839 docs/s)
Deleted 6502 docs (2816 docs/s)
Deleted 6752 docs (2724 docs/s)
⚠  firestore: delete transaction too large, reducing batch size from 250 to 25
Deleted 7042 docs (1304 docs/s)
Deleted 7067 docs (931 docs/s)
Deleted 7092 docs (477 docs/s)
Deleted 7142 docs (474 docs/s)
Deleted 7167 docs (469 docs/s)
Deleted 7192 docs (466 docs/s)
Deleted 7217 docs (386 docs/s)
Deleted 7242 docs (361 docs/s)
Deleted 7267 docs (361 docs/s)
Deleted 7292 docs (360 docs/s)
Deleted 7317 docs (357 docs/s)
Deleted 7342 docs (353 docs/s)
Deleted 7367 docs (332 docs/s)
Deleted 7368 docs (297 docs/s)
Deleted 7393 docs (298 docs/s)
Deleted 7443 docs (299 docs/s)
Deleted 7468 docs (296 docs/s)
Deleted 7543 docs (298 docs/s)
Deleted 7568 docs (290 docs/s)
Deleted 7593 docs (290 docs/s)
Deleted 7618 docs (290 docs/s)
Deleted 7643 docs (288 docs/s)
```
